contracts/cosmwasm-vm/cw-cluster-connection/src/contract.rs

```diff
@@ -0,0 +1,302 @@
+        }
+
+        self.store_signature_threshold(deps.storage, threshold)?;
+
+        Ok(Response::new().add_attribute("action", "set_validators"))
+    }
+
+    pub fn send_message(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        to: NetId,
+        sn: i64,
+        msg: Vec<u8>,
+    ) -> Result<Response, ContractError> {
+        self.ensure_xcall(deps.storage, info.sender)?;
+
+        let next_conn_sn = self.get_next_conn_sn(deps.storage)?;
+
+        let mut fee = 0;
+
+        if sn >= 0 {
+            fee = self.get_fee(deps.storage, to.clone(), sn > 0)?.into();
+        }
+
+        let value = self.get_amount_for_denom(&info.funds, self.get_denom(deps.storage));
+
+        if fee > value {
+            return Err(ContractError::InsufficientFunds);
+        }
+
+        Ok(Response::new()
+            .add_attribute("action", "send_message")
+            .add_event(
+                Event::new("Message")
+                    .add_attribute("targetNetwork", to.to_string())
+                    .add_attribute("connSn", next_conn_sn.to_string())
+                    .add_attribute("msg", self.hex_encode(msg)),
+            ))
+    }
+
+    pub fn recv_message(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        src_network: NetId,
+        conn_sn: u128,
+        msg: String,
+    ) -> Result<Response, ContractError> {
+        self.ensure_relayer(deps.storage, info.sender)?;
+
+        let vec_msg: Vec<u8> = self.hex_decode(msg)?;
+
+        if self.get_receipt(deps.as_ref().storage, src_network.clone(), conn_sn) {
+            return Err(ContractError::DuplicateMessage);
+        }
+        self.store_receipt(deps.storage, src_network.clone(), conn_sn)?;
+
+        let xcall_submessage =
+            self.call_xcall_handle_message(deps.storage, &src_network, vec_msg)?;
+
+        Ok(Response::new().add_submessage(xcall_submessage))
+    }
+
+    pub fn set_signature_threshold(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        threshold: u8,
+    ) -> Result<Response, ContractError> {
+        self.ensure_admin(deps.storage, info.sender)?;
+
+        self.store_signature_threshold(deps.storage, threshold)?;
+
+        Ok(Response::new().add_attribute("action", "set_signature_threshold"))
+    }
+
+    pub fn recv_message_with_signatures(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        src_network: NetId,
+        conn_sn: u128,
+        msg: String,
+        signatures: Vec<Vec<u8>>,
+    ) -> Result<Response, ContractError> {
+        self.ensure_relayer(deps.storage, info.sender)?;
+
+        let vec_msg: Vec<u8> = self.hex_decode(msg)?;
+
+        let threshold = self.get_signature_threshold(deps.storage);
+        let validators = self.get_validators(deps.storage)?;
+
+        self.verify_signatures(
+            deps.as_ref(),
+            threshold,
+            validators,
+            vec_msg.clone(),
```

Comment:
the signed message should be srcNetwork+sn+msg not only message.

Reply:
yep... rlp encoded the combination